src/main/java/htsjdk/samtools/util/GZIIndex.java

```diff
@@ -0,0 +1,365 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package htsjdk.samtools.util;
+
+import htsjdk.tribble.index.Index;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.ByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a .gzi index of a block-compressed file.
+ *
+ * <p>The .gzi index is a mapping between the offset of each block in the gzipped file and the
+ * uncompressed offset that that block starts with. This mapping is represented by {@link IndexEntry}.
+ *
+ * <p>An example of usage for this index for random access a bgzip file using an index generated
+ * from raw data. For example, for indexing a compressed FASTA file the .gzi index can be used in
+ * conjunction with the {@link htsjdk.samtools.reference.FastaSequenceIndex} to seek concrete
+ * sequences.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ * @see <a href="http://github.com/samtools/htslib/issues/473">https://github.com/samtools/htslib/issues/473</a>
+ */
+public final class GZIIndex {
+
+    /** Default extension for the files storing a {@link GZIIndex}. */
+    public static final String DEFAULT_EXTENSION = ".gzi";
+
+    /**
+     * Index entry mapping the block-offset (compressed offset) to the uncompressed offset where the
+     * block starts.
+     *
+     * @see <a href="http://github.com/samtools/htslib/issues/473">https://github.com/samtools/htslib/issues/473</a>
+     */
+    public static final class IndexEntry {
+        private final long compressedOffset;
+        private final long uncompressedOffset;
+
+        private IndexEntry(final long compressedOffset, final long uncompressedOffset) {
+            if (Long.signum(compressedOffset) == -1) {
```

Comment:
this might be clearer as just checks for `compressedOffset < 0`

Reply:
Changed!